File: hearts-mike/src/GameIO.hs
{-# LANGUAGE BlockArguments #-}
{-# LANGUAGE ConstraintKinds #-}
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE GADTs #-}
{-# LANGUAGE LambdaCase #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE PolyKinds #-}
{-# LANGUAGE RankNTypes #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TupleSections #-}
{-# LANGUAGE TypeApplications #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE TypeOperators #-}
{-# LANGUAGE UndecidableInstances #-}
{-# OPTIONS_GHC -fplugin=Polysemy.Plugin #-}

module GameIO where

import Data.IORef (IORef)
import qualified Data.IORef as IORef
import Polysemy
import Polysemy.Cont
import Polysemy.Internal (send)
import Polysemy.State (State)
import qualified Polysemy.State as State

import Cards
import GameEvent
import Table
import Game

gameIO :: [Player] -> IO (GameCommand -> IO [GameEvent])
gameIO players =
  do
    contRef <-
      IORef.newIORef
        ( Just
            ( \command ->
                do
                  let game = playLoop command
                  let game' = State.evalState (emptyTableState players) game
                  let gameOne =
                        callCC
                          ( \cont ->
                              fmap GameDone (Game.playOne cont game')
                          )
                  return (run (runContPure gameOne))
            )
        )
    -- loop until we get to WaitingForCommand or GameDone
    let gatherEvents' stepResult events =
          case stepResult of
            EventBroadcast event cont ->
              do
                let stepResult' = run (runContPure (cont ()))
                gatherEvents' stepResult' (events ++ [event])
            WaitingForCommand cont ->
                return (events, Just cont)
            GameDone _ -> return (events, Nothing)

    let process command =
          do
            mbCont <- IORef.readIORef contRef
            case mbCont of
              Nothing -> return []
              Just cont ->
                  do let stepResult = run (runContPure (cont command))
                     (events, mbCont') <- gatherEvents' stepResult []
                     IORef.writeIORef contRef mbCont'
                     return events
    return process

File: hearts-mike/src/main/scala/funar/hearts/GameIO.scala
package funar.hearts

import org.atnos.eff._
import org.atnos.eff.interpret._
import org.atnos.eff.all._
import org.atnos.eff.syntax.all._
import org.atnos.eff.addon.cats.effect._
import org.atnos.eff.syntax.addon.cats.effect._
import cats.Traverse
import cats.data._
import cats.implicits._
import scala.annotation.tailrec
import cats.effect.IO

object GameIO {
  import Table._ 
  import GameStepResult._

  @tailrec
  private def gatherEvents[A](stepResult: GameStepResult[A], events: Seq[GameEvent]): (Seq[GameEvent], Option[GameCommand => GameStepResult[A]]) =
    stepResult match {
      case EventBroadcast(event, cont) =>
        val stepResult1 = cont(())
        gatherEvents(stepResult1, events :+ event)
      case WaitingForCommand(cont) =>
        (events, Some(cont))
      case GameDone(_) => (events, None)
    }

  def gameIO(players: List[Player]): GameCommand => IO[Seq[GameEvent]] = {
    var ocont: Option[GameCommand => GameStepResult[Option[Player]]] =
      Some { (command: GameCommand) =>
              val game: Eff[Fx.fx2[GameStep, State[TableState, _]], Option[Player]] = Game.playLoop(command)
              val game1 = evalState(emptyTableState(players))(game)
              Game.playOne(game1)
            }
    { (command: GameCommand) =>
      ocont match {
        case None => IO.pure(Seq.empty)
        case Some(cont) => IO {
          val stepResult = cont(command)
          val (events, newCont) = gatherEvents(stepResult, Seq.empty)
          ocont = newCont
          events
        }
      }
    }
  }

}